File: Jaehan/리프노드수최대화/리프노드수최대화.py
def solution(dist_limit, split_limit):
    max_steps = 1
    while 2 ** max_steps <= split_limit:
        max_steps += 1
    max_steps += 2

    def simulate(c2, c3):
        m, p, D, leaves = 1, 1, dist_limit, 0
        for i in range(c2 + c3):
            b = 2 if i < c2 else 3      # 2를 먼저, 3을 나중에
            if D == 0:
                break
            if p * b > split_limit:      # 분배도 초과 방지
                break
            n = min(m, D)                # 가능한 만큼 최대로 분배
            leaves += m - n              # 분배 못한 노드는 리프로 확정
            m = n * b
            p = p * b
            D -= n
        leaves += m
        return leaves

    best = 0
    for c2 in range(max_steps + 1):
        for c3 in range(max_steps - c2 + 1):
            best = max(best, simulate(c2, c3))
    return best

File: Hoseong/리프노드수최대화/리프노드수최대화.cpp
#include <bits/stdc++.h>
using namespace std;

long long D, S;

map<tuple<long long,long long,long long>, long long> memo;


long long dfs(long long nodes, long long split, long long used)
{
    if (used > D || split > S)
        return 0;


    auto key = make_tuple(nodes, split, used);

    if (memo.count(key))
        return memo[key];


    // 현재 노드를 모두 리프로 종료
    long long ret = nodes;


    // 분배 노드 사용
    if (used + nodes <= D)
    {
        for (long long mul : {2LL,3LL})
        {
            long long child = nodes * mul;
            long long nextSplit = split * mul;


            if (nextSplit > S)
                continue;


            /*
                child 중 일부만 다음 분배 노드가 된다.
                나머지는 리프
            */
            for (long long next = 0; next <= child; next++)
            {
                long long leaves = child - next;

                if (next == 0)
                {
                    ret = max(ret, leaves);
                }
                else
                {
                    ret = max(
                        ret,
                        leaves + dfs(
                            next,
                            nextSplit,
                            used + nodes
                        )
                    );
                }
            }
        }
    }


    return memo[key] = ret;
}


int solution(int dist_limit, int split_limit)
{
    D = dist_limit;
    S = split_limit;

    /*
      root의 자식 1개가 최초 후보
    */
    return dfs(1,1,0);
}